docs/low-code-spark/extensibility/gem-builder.md

```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
+    diagnostics.toList
+  }
+```
+</TabItem>
+</Tabs>
+
+````
+
+### State Changes
+
+The `onChange` method is given for the UI State transformations. You are given both the previous and the new incoming state and can merge or modify the state as needed. The properties of the Gem are also accessible to this function, so functions like selecting columns, etc. are possible to add from here.
+
+````mdx-code-block
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+def onChange(self, oldState: Component[FilterProperties], newState: Component[FilterProperties]) -> Component[
+        FilterProperties]:
+        newProps = newState.properties
+        usedColExps = getColumnsToHighlight2([newProps.condition], newState)
+        return newState.bindProperties(replace(newProps, columnsSelector=usedColExps))
+
+```
+</TabItem>
+<TabItem value="scala" label="Scala">
+
+```scala
+def onChange(oldState: Component, newState: Component)(implicit context: WorkflowContext): Component = {
+    val newProps = newState.properties
+    val portId = newState.ports.inputs.head.id
+
+    val expressions = getColumnsToHighlight(List(newProps.condition), newState)
+
+    newState.copy(properties = newProps.copy(columnsSelector = expressions))
+  }
+```
+</TabItem>
+</Tabs>
+
+````
+
+### Component Code
+
+The last class used here is `FilterCode` which is inherited from `ComponentCode` class. This class contains the actual Spark code that needs to run on your Spark cluster. Here the above User Defined properties are accessible using `self.props.{property}`. The Spark code for the Gem logic is defined in the apply function.
+For example, we are calling the `.filter()` method in this example in the apply function.
+
+````mdx-code-block
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+class FilterCode(ComponentCode):
+def __init__(self, newProps):
+self.props: Filter.FilterProperties = newProps
+
+        def apply(self, spark: SparkSession, in0: DataFrame) -> DataFrame:
+            return in0.filter(self.props.condition.column())
+```
+</TabItem>
+<TabItem value="scala" label="Scala">
+
+```scala
+class FilterCode(props: PropertiesType)(implicit context: WorkflowContext) extends ComponentCode {
+
+    def apply(spark: SparkSession, in: DataFrame): DataFrame = {
+      val out = in.filter(props.condition.column)
+      out
+    }
+
+  }
+```
+</TabItem>
+</Tabs>
+
+````
+
+You can go ahead and preview the component in the Gem Builder now to see how it looks. You can modify the properties and then save it to preview the generated Spark code which will eventually run on your cluster.
+
+## DataSources Gems
+
+DataSource Gems (also referred to as Source/Target Gems) are Gems that you use to read/write your Datasets into DataFrames. The only major difference between a Source/Target Gem and a Transformation Gem is that a Source/Target Gem will have two `dialog` and two `apply` functions each for Source and Target respectively. Let's look at them with an example
```

Comment:
In this paragraph there's only one major difference, but at the end we see there are three differences listed. 
I suggest to be consistent - mention all three differences in both the intro and the summary at the end.